network-api/networkapi/buyersguide/tests.py

```diff
@@ -440,18 +460,455 @@ def name_change_changes_slug(self):
+        """
+        software_products = SoftwareProduct.objects.all()
+
+        for product in software_products:
+            # Find the equiv ProductPage.
+            product_page = ProductPage.objects.get(slug=product.slug).specific
+            # Get all the General Product Page fields.
+            specific_fields = self.general_product_fields + [
+                'medical_privacy_compliant', 'easy_to_learn_and_use', 'handles_recordings_how',
+                'recording_alert', 'recording_alert_helptext', 'medical_privacy_compliant_helptext',
+                'host_controls', 'easy_to_learn_and_use_helptext'
+            ]
+            # For every field, compare it to the old field in the Product object
+            for field in specific_fields:
+                self.assertEqual(
+                    getattr(product, field),
+                    getattr(product_page, field),
+                    f"{field} did not match on product: {product} (id:{product.id})",
+                )
+
+    def check_general_products_match(self):
+        """
+        Compare all GeneralProducts with all GeneralProductPages and all their fields
+        """
+        general_products = GeneralProduct.objects.all()
+        for product in general_products:
+            # Find the equiv ProductPage.
+            product_page = ProductPage.objects.get(slug=product.slug).specific
+            # Get all the General Product Page fields.
+            specific_fields = self.general_product_fields + [
+                'camera_device', 'camera_app', 'microphone_device', 'microphone_app',
+                'location_device', 'location_app', 'personal_data_collected',
+                'biometric_data_collected', 'social_data_collected',
+                'how_can_you_control_your_data', 'data_control_policy_is_bad',
+                'track_record_choices', 'company_track_record', 'track_record_is_bad',
+                'track_record_details', 'offline_capable', 'offline_use_description',
+                'uses_ai', 'ai_uses_personal_data', 'ai_is_transparent', 'ai_helptext'
+            ]
+            # For every field, compare it to the old field in the Product object
+            for field in specific_fields:
+                self.assertEqual(
+                    getattr(product, field),
+                    getattr(product_page, field),
+                    f"{field} did not match on product: {product} (id:{product.id})",
+                )
+
+    def test_migration(self):
+        """
+        All tests go in here to ensure they are executed in the proper order
+        """
+        self.before_migrate()
+        self.migrate()
+        self.after_migrate()
+        self.check_buyersguide_exists()
+        self.check_general_products_match()
+        self.check_software_products_match()
+
+
+class TestProductPage(BuyersGuideTestMixin):
+
+    def setUp(self):
+        super().setUp()
+        if not hasattr(self.product_page.votes, 'get_votes'):
```

Comment:
I'm not sure about the logic here: if `self.product_page.votes` isn't `None`, can it be anything other than our votes object? Should this be `if not self.product_page.votes: ...` instead?

Reply:
No it cant be anything other than `None` or a `ProductPageVotes()` object. This is checking to see if there is a Votes object and it's not None. But this test requires this little quirk you mentioned:
```python
self.assertFalse(hasattr(self.product_page.votes, 'set_votes'))
```

What this is doing is checking that the Votes object is not attached. Honestly, not sure why your suggestion doesnt work - it reads like it should.